### DISCUSSION TITLE
How to add a text to the chart

### DISCUSSION THREAD

**Discussion by asmirn1** [GSI-CS-CO/chart-fx#448] [Q&A] [answered]
Hi,

I'd like to add some text to a chart. Currently, I was able do it using `XValueIndicator`. Here is my code:
```java
XValueIndicator indicator = new XValueIndicator(axis, mz, text);
indicator.setLabelPosition(relativeHeight);
indicator.setLabelVerticalAnchor(relativeHeight < 0.5 ? VPos.BOTTOM : VPos.TOP);
indicator.setEditable(false);
chart.getPlugins().add(indicator);
```

I've got the following result:
<img width="1243" alt="Screen Shot 2021-07-06 at 2 11 18 PM" src="https://user-images.githubusercontent.com/13631120/124647996-acac4900-de64-11eb-990d-d4f4a71e7e4c.png">

However, I'd like to do several changes:
- Remove the vertical dashed lines, and
- Stop users from moving the text (currently, users can drag the text anywhere along the x-axis)

I want the result to look something like this:
<img width="1241" alt="Screen Shot 2021-07-06 at 1 51 35 PM" src="https://user-images.githubusercontent.com/13631120/124648500-525fb800-de65-11eb-99ba-50bd1ce72536.png">

Please could you explain to remove the vertical lines and make the text static.

Also, is there any other way to add a text on a chart? Can I implement this using `dataSet.addDataLabel()`? I couldn't understand how.

Thank you

**Comment by wirew0rm** [comment, marked as answer]
`setEditable(false)` seems to only affect dragging on the line and not on the label, this is a bug, which we should fix.
For not drawing the line, you could use css, check the AbstractSingleValueIndicatorSample for an example of using css to style the indicator.

On the other hand, if you only want to have text, you could also implement your own plugin (probably based on one of the ValueIndicator variants). This should be pretty simple and would give you a lot more influence on how and where everything is rendered.

Looking at your screenshot, you seem to want to add labels to data points, for that usecase there is also the `LabeledMarkerRenderer`, which allows you to add your strings into your dataset. You would also have to suppress the drawing of the lines and possibly customize the renderer to put the labels at the right positions. Compared to the Plugin approach, this uses the grapics context instead of the scene graph and is non interactive by default.

**Reply by asmirn1** [reply]
@wirew0rm Thank you for a detailed explanations.

I was able to create stylesheet for my chart:
```css
.x-value-indicator-line0 { 
    -fx-stroke-width: 0; 
}
```
and the result looks like this:
<img width="1241" alt="Screen Shot 2021-07-07 at 9 30 17 AM" src="https://user-images.githubusercontent.com/13631120/124767726-12044680-df06-11eb-8ac4-3c219b4fee1a.png">

I've noticed that this style is applied only to the first line (because of the index 0 in `.x-value-indicator-line0` as I understand). How can I apply this style to all indicator lines if I don't know the exact number of indicator lines beforehand?

**Reply by wirew0rm** [reply]
just leave out the trailing zero. Also the issue with the indicator not respecting set editable(false) should be fixed once #446 gets merged.

**Reply by asmirn1** [reply]
It works now. Thank you!

**Reply by RalphSteinhagen** [reply]
@asmirn1 and @wirew0rm just merged the fix.

**Reply by RalphSteinhagen** [reply]
@asmirn1 did this answer/solve your issue?

**Reply by asmirn1** [reply]
@RalphSteinhagen Yes, it did.

I'm not sure how to close a discussion, so feel free to close it if it's possible.

**Reply by RalphSteinhagen** [reply]
@asmirn1 you could select @wirew0rm reply as an 'answer' that'll help. 

Thanks again for your interest in chart-fx. Hope this helps in your purposes...
